libs/wingc/src/lib.rs

```diff
@@ -286,15 +283,42 @@ pub fn type_check(
 	tc.type_check_file_or_dir(file, scope);
 }
 
-pub fn compile(
-	project_dir: &Utf8Path,
-	source_path: &Utf8Path,
-	source_text: Option<String>,
-	out_dir: &Utf8Path,
-) -> Result<CompilerOutput, ()> {
-	let source_package = as_wing_library(project_dir).unwrap_or_else(|| DEFAULT_PACKAGE_NAME.to_string());
+/// Infer the root directory of the current Wing application or library.
+///
+/// Check the current file's directory for a wing.toml file or package.json file,
+/// and continue searching up the directory tree until we find one.
+/// If we run out of parent directories, fall back to the first directory we found.
+pub fn find_nearest_wing_project_dir(source_path: &Utf8Path) -> Utf8PathBuf {
+	let initial_dir = if source_path.is_dir() {
+		source_path.to_owned()
+	} else {
+		source_path.parent().unwrap().to_owned()
+	};
+	let mut current_dir = initial_dir.as_path();
+	loop {
+		if current_dir.join("wing.toml").exists() {
+			return current_dir.to_owned();
+		}
+		if current_dir.join("package.json").exists() {
+			return current_dir.to_owned();
+		}
+		if current_dir == "/" {
+			break;
+		}
+		current_dir = current_dir.parent().unwrap();
+	}
+	return initial_dir;
+}
+
+pub fn compile(source_path: &Utf8Path, source_text: Option<String>, out_dir: &Utf8Path) -> Result<CompilerOutput, ()> {
+	let project_dir = find_nearest_wing_project_dir(source_path);
```

Comment:
I updated the main `compile` function so it's now responsible for looking through your file system for the root project you're in. The main motivation is to unify the logic with the LSP -- in the LSP, we don't know where the root of the project is, which makes it tricky for us to know where is the "root" is for the purposes of automatically calculating FQNs according to the user's directory structure. So in both places, we now try to look for a package.json or wing.toml and if we find one, we treat that as our project root.

There's a chance this heuristic will be too aggressive 🤷 - but I think the best way to go about this is eventually to lean into it and to recommend building projects with `wing.toml` files.